x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationEngine.java

```diff
@@ -0,0 +1,111 @@
+
+    void resolveAuthorizationInfo(Authentication authentication, TransportRequest request, String action,
+                                  ActionListener<AuthorizationInfo> listener);
+
+    void authorizeRunAs(Authentication authentication, TransportRequest request, String action, AuthorizationInfo authorizationInfo,
+                        ActionListener<AuthorizationResult> listener);
+
+    void authorizeClusterAction(Authentication authentication, TransportRequest request, String action, AuthorizationInfo authorizationInfo,
+                                ActionListener<AuthorizationResult> listener);
+
+    boolean checkSameUserPermissions(String action, TransportRequest request, Authentication authentication);
+
+    boolean shouldAuthorizeIndexActionNameOnly(String action, TransportRequest request);
+
+    void authorizeIndexActionName(Authentication authentication, TransportRequest request, String action,
+                                  AuthorizationInfo authorizationInfo, ActionListener<AuthorizationResult> listener);
+
+    List<String> loadAuthorizedIndices(Authentication authentication, TransportRequest request, String action, AuthorizationInfo info,
+                                       Map<String, AliasOrIndex> aliasAndIndexLookup);
+
+    void buildIndicesAccessControl(Authentication authentication, TransportRequest request, String action,
+                                   AuthorizationInfo authorizationInfo, Set<String> indices,
+                                   SortedMap<String, AliasOrIndex> aliasAndIndexLookup, ActionListener<IndexAuthorizationResult> listener);
+
+    interface AuthorizationInfo {
+
+        Map<String, Object> asMap();
+
+        default AuthorizationInfo getAuthenticatedUserAuthorizationInfo() {
+            return this;
+        }
+    }
+
+    final class EmptyAuthorizationInfo implements AuthorizationInfo {
+
+        public static final EmptyAuthorizationInfo INSTANCE = new EmptyAuthorizationInfo();
+
+        private EmptyAuthorizationInfo() {}
+
+        @Override
+        public Map<String, Object> asMap() {
+            return Collections.emptyMap();
+        }
+    }
+
+    class AuthorizationResult {
+
+        private final boolean granted;
+        private final boolean auditable;
+
+        public AuthorizationResult(boolean granted) {
+            this(granted, true);
+        }
+
+        public AuthorizationResult(boolean granted, boolean auditable) {
+            this.granted = granted;
+            this.auditable = auditable;
+        }
+
+        public boolean isGranted() {
+            return granted;
+        }
+
+        public boolean isAuditable() {
+            return auditable;
+        }
+
+        public static AuthorizationResult granted() {
+            return new AuthorizationResult(true);
+        }
+
+        public static AuthorizationResult deny() {
+            return new AuthorizationResult(false);
+        }
+    }
```

Comment:
I see where you are coming from. There are four distinct cases, so we'd have something like

```java
public enum AuthorizationResult {

    GRANTED_AUDIT(true, true),
    GRANTED_DONT_AUDIT(true, false),
    DENIED_AUDIT(false, true),
    DENIED_DONT_AUDIT(false, false);

    private final boolean granted;
    private final boolean auditable;

    public AuthorizationResult(boolean granted, boolean auditable) {
        this.granted = granted;
        this.auditable = auditable;
    }

    public boolean isGranted() {
        return granted;
    }

    public abstract boolean isAuditable() {
        return auditable;
    }
}
```

Is that what you're thinking?

So one thing missing from this PR, but which was in my initial proposal is a `details` or `reason` field, https://github.com/elastic/elasticsearch/issues/32435#issuecomment-408445882. This field could be used to explain why the authorization result was made. Eventually the plan was to propagate that through to the audit log. I think this has value, but what do you think?

Reply:
> Is that what you're thinking?

yes.

> Eventually the plan was to propagate that through to the audit log. I think this has value, but what do you think?

Okay, I suppose we have more leeway if we keep it a class rather than an enum.